### PR TITLE
implement `@expect` builtin

### LINKED ISSUE #489: @expect() hint to optimizer
This is proposal for a small feature, local and isolated. Could improve code readability.

------------
Linux kernel uses macros `likely` and `unlikely` :

```
if (likely(x > 0)) { ... }
...
if (unlikely(y == NULL)) { ... }

```

These macros may improve performance a little bit and they serve as handy documentation.

Language Nim supports this functionality too, through its standard library ( https://nim-lang.org/docs/system.html#likely.t,bool ).

----------------------

**My proposal:**

Add `if+` and `if-` into the language. `if+` would signal that the path is rather likely, `if-` will suggest the flow will not go this way.

```
if+ (x < 0) { 
  ...  // likely goes here
}
if- (err) { 
  ... // unlikely
}

```

What is this good for:
1. It gives hint to source code reader, hint which will be almost never found in the documentation. This hint is very short, intuitive, and doesn't require extra pair of parenthesis.
2. It may slightly improve the performance, by rearranging instructions so that the likely path goes w/o jump, or by inserting hint for branch selector. 

How it could be implemented:
1. By doing nothing, just using it as documentation only feature.
2. By using IR opcode `llvm.expect`. This is how clang implements `__builtin_expect`, which is called by `likely`/`unlikely `macros. (GCC also provides `__builtin_expect`, MSVC has nothing similar.)

-------------------

Performance effects of `__builtin_expect` are hotly disputed on the internet. Many people claim programmers are invariably bad in such prediction and profile guided optimization will do much, much better. However, they never support their claim by benchmarks.

Even if performance is unaffected the documentation value remains. When one is stepping the code through debugger and the flow goes against the hint, one may be get more cautious a catch a bug.

### PR DESCRIPTION
**Pull request by Rexicon226**
closes #489 

### Why `@expect` takes a bool:
- `@expect` affects branching codegen and you can only branch on a bool.
- Expecting a specific value can be done with `unreachable` with respect to the optimizer.

### Why `@expect` doesn't have a probability:
- Probabilities don't make a practical difference in codegen vs. binary.
- The switch usecase can't reliably predict codegen (could be a jump table, some could be comparisons, or it could be a mixture of both)

### CODE REVIEW

**Comment by daurnimator** [comment]
> Probabilities don't make a practical difference in codegen vs. binary.

Are you saying that LLVM's `expect.with.probability` doesn't do anything?

**Comment by kprotty** [comment]
Expect with probably seems to affect `switch` cases but it's unclear how it does so practically compared to `if`.

**Comment by Techatrix** [comment]
When there is no expected value nor probability, wouldn't a name like `@likely` instead of `@expect` be more fitting so that this builtin doesn't get mixed up to be an assertion or function like `std.testing.expect`?

**Comment by notcancername** [comment]
For purposes of consistency with C (compiler `__builtin_expect` vs the common `#define LIKELY(x) __builtin_expect()`), it seems better to name it `likely`.

**Comment by Rexicon226** [comment]
I don't want likely as it implies an unlikely. Note that the only real purpose of this builtin *is* the unlikely version, and unlikely is a bit ugly imo. 

I will change it up to accept a second field for true or false instead and keep expect as per @kprotty suggestion when we talked about it.

**Comment by notcancername** [comment]
Since the user can implement a likely or unlikely pretty easily themselves when a second field is added:
```
inline fn likely(x: bool) bool {
    return @expect(x, true);
}
inline fn unlikely(x: bool) bool {
    return @expect(x, false);
}
```
I think this would be the best solution. Thanks! :)

**Comment by notcancername** [comment]
Should `likely` and `unlikely` functions be part of the standard library? On one hand, they're one-liners any user can easily write, on the other hand, everyone writing their `likely` and `unlikely` functions is worse, a likely/unlikely micropackage would also suck.

**Comment by Rexicon226** [comment]
Thought about that yesterday, but I think making `@expect` take a second "expected" bool supports both `likely` and `unlikely` without a wrapper.

**Review by clickingbuttons — COMMENTED** [review]
What about a nice sugary `@expect(operand: anytype, expected: anytype)` and add a check for `@typeOf(operand) == @typeOf(expected)`?

File: doc/langref.html.in [hunk 1]
```diff
@@ -8726,6 +8726,24 @@ export fn @"A function name that is a complete sentence."() void {}
       {#see_also|@export#}
       {#header_close#}
 
+      {#header_open|@expect#}
+      <pre>{#syntax#}@expect(operand: bool, expected: bool) bool{#endsyntax#}</pre>
+      <p>
+      Informs the optimizer that {#syntax#}operand{#endsyntax#} will likely be {#syntax#}expected{#endsyntax#}, which influences branch compilation to prefer generating the true branch first.
```
Comment (kprotty) [inline doc/langref.html.in:8732]: What about
> Returns a new bool with the value of `operand` which informs the compiler that it's likely equal to `expected`. The compiler may use this information to improve branch generation by preferring to emit the most likely branch first.

Also add **comptime** `expected: bool`

**Review by kprotty — APPROVED** [review]

**Comment by Snektron** [comment]
image
nobody `@expect`s the Spanish inquisition!!

**Comment by andrewrk** [comment]
Reverted in 9be8a9000faead40b1aec4877506ff10b066659c. I would like a chance to review this please.

**Pull request merged**